plugins/coverimage.koplugin/main.lua

```diff
@@ -0,0 +1,268 @@
+local Device = require("device")
+
+if not Device.isAndroid() and not Device.isEmulator() then
+    return { disabled = true }
+end
+
+local InfoMessage = require("ui/widget/infomessage")
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local ffiutil = require("ffi/util")
+local lfs = require("libs/libkoreader-lfs")
+local logger = require("logger")
+local util = require("util")
+local _ = require("gettext")
+local T = require("ffi/util").template
+
+local function pathOk(filename)
+    local path, name = util.splitFilePathName(filename)
+    return Device:isValidPath(path) and name ~= "" and lfs.attributes(filename, "mode") ~= "directory"
+end
```

Comment:
You never answered why this exists.

Reply:
This is for safety, so that we do not overwrite an existing file somewhere outside the sandbox.

1.) We could drop the `isValidPath` check (maybe for flavor "rocks"):
  * Bonus: We could overwrite a lot of image files (sandbox + all in KOReader app directory + a lot of in `/data/data`).
  * Malus: We could write to places, where we should not (e.g. other apps).
  * If we drop `isValidPath` then we should at least check if the extension is `png, jpg, jpeg`.

2.) We have to check if the cover image filename is not a directory, otherwise `copyFile` crashes.

3.) We have to check that the path does not end with a slash, otherwise `copyFile` crashes.

Reply:
So something closer to:

```suggestion
local function pathOk(filename)
    return Device:isValidPath(filename) and util.fileExists(filename)
end
```

Reply:
@Frenzie: Hmm, your suggested solution is not right. If the file exists, it gets overwritten. This should happen on opening a new book. 

But we need to check:
1.) Is the file on a valid path -> isValidPath().
2.) Does the path to the file really exist -> util.pathExists()
3.) If the user enters a path with an empty filename -> name~=""
4.) Has the user entered a directoryname instead of a filename -> lfs.attributes()

So we can catch with 1.) `/data/data/cover.jpg` not valid
2.) `/sdcard/kor/cover.jpg`, if `kor` does not exist
3.) `/sdcard/` empty filename
4.) `/sdcard/koreader` directoryname

In each of these cases we should tell the user, to correct the image filename.

Reply:
> Hmm, your suggested solution is not right.

It's not meant to be right; it's meant to look right. Slight difference. ;-)

> So we can catch with 1.) /data/data/cover.jpg not valid
2.) /sdcard/kor/cover.jpg, if kor does not exist
3.) /sdcard/ empty filename
4.) /sdcard/koreader directoryname

And there's the rub. That's not good UX, which explains why the function's suspicious to me. It's doing either too much or too little. The above was from the too much perspective, the following general idea from too little.

```suggestion
local function pathOk(filename)
    local path, name = util.splitFilePathName(filename)
    if not Device:isValidPath(path) then
        return false, _("Path isn't in a writable location.")
    elseif name == ""
        return false, _("Please enter a filename at the end of the path.")
    elseif lfs.attributes(filename, "mode") == "directory"
        return false, _("The path must point to a file, but it points to a directory.")
    end
    return true
end
```

The in between doesn't sit well with me.